scripts/pylib/twister/twisterlib.py

```diff
@@ -2045,6 +2050,54 @@ def fill_results_by_status(self):
         for k in self.results:
             self.results[k] = status_to_verdict[self.status]
 
+    def create_build_dir_shortcut(self):
+        """
+        Create build directory with original "long" path. Next take shorter
+        path (based on TestInstance class internal counter) and link them with
+        original path - create shortcut. At the end replace build_dir to
+        shorter path. This action helps to limit path length which can be
+        significant during building by CMake on Windows OS.
+        """
+        if os.name != "nt":  # if OS is not Windows
+            logger.warning("Creating shorter build paths available only on "
+                           "Windows OS")
+            return
+
+        os.makedirs(self.build_dir)
+
+        shortcuts_dir_name = "scs"  # folder for all shortcuts
+        shortcuts_dir_path = os.path.join(self.outdir, shortcuts_dir_name)
+        if not os.path.exists(shortcuts_dir_path):
+            os.mkdir(shortcuts_dir_path)
+
+        shortcut_created_flag = False
+        # this checking in while loop is performed in case when Twister is run
+        # in several processes in pararel and to avoid overwriting the
+        # same shortcut_name by two instances
+        while not shortcut_created_flag:
+            shortcut_name = self.get_new_shortcut_name(shortcuts_dir_path)
+            shortcut_path = os.path.join(shortcuts_dir_path, shortcut_name)
+            command = ["mklink", "/J", f"{shortcut_path}", f"{self.build_dir}"]
+            result = subprocess.call(command, shell=True)
+            if result == 0:
+                shortcut_created_flag = True
```

Comment:
I don't think running multiple twister instances is a real use case as it doesn't bring any value, since twister already handles parallelization internally. Also, each instance will create its own output dir, so if shortcuts are there, there won't be any conflict between shortcuts from multiple instances.  Running multiple twister instances for subset is AFAIK used either in parallel on separate agents or on a single agent sequentially.
I am not sure if the proposed mechanism will prevent race conditions which might cause mess in links when build processes run in parallel within a single twister instance. 
What do you think about using locks? E.g. the counter variable is define at the twister initialization and has getter/setter methods with a lock. When a shortcut is created, the variable is locked, the counter is increased, shortcut created, and the lock is released? Using such approach might also optimize the code a bit, since you won't need to scan the whole catalog looking for the last number and instead get it from the counter immediately. You can check here how it was implemented for counting test cases: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib.py#L94